### PR TITLE
Add Uptime Kuma checks for built-site maintenance

### PR DESCRIPTION
**Pull request by stefan-burke**
## Summary

- show the scheduled maintenance check on each built site maintenance tab
- let an owner add a missing authenticated `POST /scheduled` check
- support Uptime Kuma 2.4 or newer through the runtime native WebSocket
- reuse existing checks when their method, URL, bearer authorization, and accepted status range are equivalent
- show paused checks as paused instead of creating duplicates
- keep concurrent additions safe without deleting shared groups
- translate local connection, protocol, timeout, version, and credential failures through the message catalog
- preserve plain error messages returned by Uptime Kuma

## Operator setup

Set `CAN_BUILD_SITES=true`, `UPTIME_KUMA_URL`, `UPTIME_KUMA_USERNAME`, and `UPTIME_KUMA_PASSWORD` on the builder. The Uptime Kuma server must be version 2.4 or newer. `UPTIME_KUMA_INTERVAL_MINUTES` is optional, accepts any positive whole number, and defaults to 15.

The maintenance tab reports whether a check exists and offers an add action when it is missing. Each new check uses Kuma built-in bearer authentication. Request headers and bearer tokens are not added to rendered page state.

Accounts that require two-factor login are not supported. A paused check is shown to the operator and must be resumed in Uptime Kuma.

## Reliability

The client logs in before waiting for the version that Kuma reveals after login. It validates the version before reading or writing checks. Incorrect credentials produce clear builder-facing catalog copy instead of Kuma internal translation keys.

Monitor-list events stay open until the matching request acknowledgement, so an earlier snapshot cannot replace the requested list. Large monitor lists may take up to 60 seconds. Server disconnects clear connection state immediately.

Duplicate shared groups are supported, and new checks use the lowest group number. Concurrent add requests reuse the lowest winning check and delete only a duplicate check created by that same request. Shared groups are never deleted.

## Testing

- pinned Deno 2.5.6 `deno task precommit` passes after merging current `main`
- direct regressions cover login ordering, the Kuma 2.4 minimum, incorrect credentials, two-factor login, connection and protocol failures, large monitor lists, disconnect state, equivalent authentication and URLs, accepted status ranges, paused checks, concurrent creation, and catalog-backed operator copy
- exhaustive mutation: Kuma socket 236/236, client 262/262, monitor reconciliation 212/212, and shared Kuma error 1/1
- exhaustive mutation: shared parser 7/7 and stored JSON 10/10
- full line and branch coverage passes

### CODE REVIEW

File: src/shared/uptime-kuma/monitors.ts [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+const siteMonitor = (
+  monitors: UptimeKumaMonitor[],
+  groupId: number,
+  url: string,
+): UptimeKumaMonitor | null => {
+  const matches = monitors.filter(
+    (monitor) =>
+      monitor.type === "http" &&
+      monitor.parent === groupId &&
+      monitor.url === url,
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in d43b36ea. The shared lookup now requires `method === "POST"`, so a GET check stays visible to Kuma but no longer hides or blocks the required scheduled-maintenance monitor. Regression tests prove both the missing page state and creation of a new POST monitor when only a GET check exists; the updated module still has a 100% mutation score.

File: test/shared/uptime-kuma/config.test.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      CAN_BUILD_SITES: "true",
+      UPTIME_KUMA_INTERVAL_MINUTES: undefined,
+      UPTIME_KUMA_PASSWORD: undefined,
+      UPTIME_KUMA_URL: undefined,
+      UPTIME_KUMA_USERNAME: undefined,
+    });
+
+    expect(getUptimeKumaConfigOrNull()).toBeNull();
+    expect(getEnabledUptimeKumaConfigOrNull()).toBeNull();
+  });
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/uptime-kuma/config.test.ts:29]: Resolved in d6e62c87. The configuration suite now sets only `UPTIME_KUMA_INTERVAL_MINUTES` and proves that both configuration accessors reject the partial setup with the required-credentials error. The configuration module retains a 100% mutation score (59/59).

File: src/shared/uptime-kuma/monitors.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+          existingGroup === null
+            ? await client.addMonitor(groupInput())
+            : existingGroup.id;
+        const monitorId = await client.addMonitor(
+          siteInput(site, config, groupId, scheduledTaskKey),
+        );
+        return okResult({ created: true, monitorId });
+      });
+    },
+  );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts:339]: Resolved in d6e62c87. Add now re-reads after group and monitor creation, selects the lowest existing ID as the race winner, and deletes only the losing group or monitor created by that same request. Table-driven regression cases cover concurrent group creation, concurrent monitor creation, occupied groups, missing acknowledgements, and cleanup; the monitor module kills all 122 mutants.

File: src/shared/uptime-kuma/monitors.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  monitors: UptimeKumaMonitor[],
+  groupId: number,
+  url: string,
+): UptimeKumaMonitor | null => {
+  const matches = monitors.filter(
+    (monitor) =>
+      monitor.type === "http" &&
+      monitor.parent === groupId &&
+      monitor.method === "POST" &&
+      monitor.url === url,
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in e0a217a2. Monitor-list headers are validated at the gateway and kept only in the internal monitor model; reuse now parses them and requires the current site’s bearer value, with case-insensitive header names. Missing, wrong, and malformed headers remain non-matches, while rendered monitor details still contain no headers. Regression tests cover load and add behavior, and monitor reconciliation kills all 118 mutants.

File: src/shared/uptime-kuma/monitors.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  } catch (error) {
+    return failed(error);
+  }
+};
+
+const monitorDefaults = (
+  type: "group" | "http",
+  name: string,
+  parent: number | null,
+): UptimeKumaMonitorInput => ({
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in e0a217a2. Monitor input construction now includes the four JSON values that the Uptime Kuma 2 add handler stringifies: `conditions: []`, `rabbitmqNodes: []`, `kafkaProducerBrokers: []`, and `kafkaProducerSaslOptions: { mechanism: "None" }`. The input builder has an exact direct regression test and kills all 25 mutants; the service test also verifies those fields reach the add request.

File: src/shared/uptime-kuma/monitors.ts [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+  if (createdMonitorExists) await client.deleteMonitor(monitorId);
+  await deleteOwnedGroupIfEmpty(
+    client,
+    monitors,
+    ownedGroupId,
+    winner.parent,
+    monitorId,
+  );
+  return { created, monitorId: winner.id };
+};
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts:237]: Resolved in a2b87205. `raceWinnerMonitor` now owns the shared-group ID lookup and authenticated `siteMonitor` call, and both add paths use it. CPD reports 0% duplication and the monitor service kills all 117 mutants.

File: test/shared/uptime-kuma/client/protocol.test.ts [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+  test("names a blocked Kuma connection in the subrequest error", async () => {
+    await expect(
+      withSubrequestAllowance({ database: 50, external: 0, total: 50 }, () =>
+        uptimeKumaSocketFactory.create(config),
+      ),
+    ).rejects.toThrow(
+      "Blocked external operation: Uptime Kuma socket connection",
+    );
+  });
+});
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/uptime-kuma/client/protocol.test.ts:397]: Resolved in a2b87205. The client tests now live in focused protocol and connection suites with one shared fake socket under `test/shared/uptime-kuma/client/`; no test file exceeds the project target. The split suite passes directly and kills all 72 client mutants.

[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/uptime-kuma/client/protocol.test.ts:397]: Resolved in a2b87205. The client tests now live in focused protocol and connection suites with one shared fake socket under `test/shared/uptime-kuma/client/`; no test file exceeds the project target. The split suite passes directly and kills all 72 client mutants.

File: src/shared/uptime-kuma/monitors.ts [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+  const groups = sharedGroups(await client.getMonitors());
+  const group = firstById(
+    groups,
+    `Uptime Kuma did not return the new "${UPTIME_KUMA_GROUP_NAME}" group.`,
+  );
+  const createdGroupStillExists = groups.some(
+    (candidate) => candidate.id === createdGroupId,
+  );
+  if (group.id !== createdGroupId && createdGroupStillExists) {
+    await client.deleteMonitor(createdGroupId);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in a2b87205. Group reconciliation now passes the refreshed full monitor list through `deleteOwnedGroupIfEmpty`, so a newly created group is deleted only when no monitor uses it. A regression reproduces an older shared group appearing after another request has already added a monitor to this request’s group, and verifies that no delete is sent; a second case covers a group Kuma no longer returns. The monitor service kills all 117 mutants.

File: src/shared/uptime-kuma/monitors.ts [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+): Promise<void> => {
+  if (
+    ownedGroupId !== null &&
+    ownedGroupId !== keptGroupId &&
+    !monitors.some(
+      (monitor) =>
+        monitor.parent === ownedGroupId && monitor.id !== ignoredMonitorId,
+    )
+  ) {
+    await client.deleteMonitor(ownedGroupId);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in 972d7bc6 by eliminating the unsafe operation rather than narrowing its race window. Uptime Kuma has no atomic conditional group delete, so the service no longer deletes shared groups at all. Duplicate shared groups are now a supported collection: lookups span all of them and new monitors use the lowest group ID. Only a losing monitor created by the current request may be deleted. The regression injects another request attaching a monitor immediately before any attempted group delete and proves the add succeeds without issuing that delete; the monitor service kills all 97 mutants.

**Comment by stefan-burke** [comment]
@coderabbitai review

**Comment by stefan-burke** [comment]
@codex review

File: src/shared/uptime-kuma/monitors.ts [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+  allowRaceDuplicates: boolean,
+): UptimeKumaMonitor | null => {
+  const matches = monitors.filter(
+    (monitor) =>
+      monitor.type === "http" &&
+      monitor.parent !== null &&
+      groupIds.includes(monitor.parent) &&
+      monitor.method === "POST" &&
+      monitor.url === url &&
+      hasAuthorization(monitor.headers, authorization),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in e6dd71c4. The client now validates and maps `accepted_statuscodes` at the Kuma boundary, and monitor reuse requires either Kuma’s `200-299` range or exact `204`. A regression reproduces a correctly authenticated POST check that accepts only 200 and proves it remains missing; a second proves exact 204 is reusable. Monitor reconciliation kills all 101 mutants.

File: src/shared/uptime-kuma/monitor-input.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+const monitorDefaults = (
+  type: "group" | "http",
+  name: string,
+  parent: number | null,
+): UptimeKumaMonitorInput => ({
+  accepted_statuscodes: ["200-299"],
+  authMethod: "",
+  body: null,
+  conditions: [],
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitor-input.ts:20]: Resolved in e6dd71c4. The client captures Kuma’s version info before connecting so it cannot miss the post-login event, validates the version, and exposes its major number. Monitor input uses one declarative defaults table: 1.x omits `conditions` and `rabbitmqNodes`, while 2.x and later include them; the Kafka fields remain on both versions. Direct and service regressions cover both payloads. The client kills 104/104 mutants and monitor input kills 30/30.

**Comment by stefan-burke** [comment]
@coderabbitai review

**Comment by stefan-burke** [comment]
@codex review

File: src/shared/uptime-kuma/socket.ts [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+  }
+
+  #closed(): void {
+    this.#clearConnectTimer();
+    const error = new Error("Uptime Kuma connection closed.");
+    this.#rejectAcks(error);
+    if (!this.#closedByClient && !this.#connected) {
+      this.#emit("connect_error", error);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/socket.ts:198]: Resolved in b3692d6c. Close and failure paths now clear the live connection flag, while a separate ever-connected flag prevents a follow-up raw WebSocket close from being misreported as an initial connection failure. Direct regressions cover new calls after server close and protocol failure, plus calls during an incomplete handshake; the socket module kills all 141 mutants.

File: src/shared/uptime-kuma/monitor-input.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+export const scheduledAuthorization = (scheduledTaskKey: string): string =>
+  `Bearer ${scheduledTaskKey}`;
+
+type KumaGeneration = "one" | "two";
+
+const VERSION_DEFAULTS: Record<KumaGeneration, UptimeKumaMonitorInput> = {
+  one: {},
+  two: { conditions: [], rabbitmqNodes: [] },
+};
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitor-input.ts]: Resolved in b3692d6c. The version defaults table now stores factories, so every Kuma 2 monitor gets fresh conditions and rabbitmqNodes arrays. A direct identity regression proves two generated inputs do not share either array, and monitor input kills all 30 mutants.

File: src/shared/uptime-kuma/monitors.ts [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+const load = (site: BuiltSite): Promise<UptimeKumaMonitorState> =>
+  withConfiguredKuma<UptimeKumaMonitorState>(
+    () => ({ kind: "unconfigured" }),
+    (error) => ({ error: errorMessage(error), kind: "error" }),
+    (config) => loadConfigured(site, config),
+  );
+
+const add = (site: BuiltSite): Promise<Result<AddedMonitor>> =>
+  withConfiguredKuma<Result<AddedMonitor>>(
+    () => errorResult("Uptime Kuma is not configured."),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in b3692d6c. Both locally produced add failures now come from built-sites catalog keys: the unconfigured message has its own key and the missing-key path reuses built_sites.kuma_needs_key. The service tests assert those catalog values directly, and the full i18n/precommit gate passes.

File: src/shared/uptime-kuma/client.ts [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+      version.cancel();
+      socket.disconnect();
+    },
+    getMajorVersion: (): Promise<number> => version.read(),
+    getMonitors: async (): Promise<UptimeKumaMonitor[]> => {
+      const list = waitForEvent(socket, "monitorList");
+      try {
+        const [response, monitorList] = await Promise.all([
+          call(socket, "getMonitorList"),
+          list.promise,
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts]: Resolved in b3692d6c. Monitor-list reads now keep capturing monitorList events until the getMonitorList acknowledgement and use the newest captured list, so a delayed login snapshot cannot win. The regression sends a stale list and then the request list before the acknowledgement and proves the request list is returned; the client kills all 106 mutants.

File: src/shared/uptime-kuma/client.ts [hunk 16]
```diff
@@ -0,0 +1,267 @@
+import * as v from "valibot";
+import { countExternalSubrequest } from "#shared/subrequest-budget.ts";
+import { integerAtLeast } from "#shared/validation/number.ts";
+import type { UptimeKumaConfig } from "./config.ts";
+import {
+  createUptimeKumaSocket,
+  type UptimeKumaSocket,
+  uptimeKumaConnectionError,
+} from "./socket.ts";
+
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts:13]: Resolved in b3692d6c. Monitor-list reads have a dedicated 60-second timeout applied to both the getMonitorList acknowledgement and the pushed list, while ordinary socket calls remain at 10 seconds. A fake-time regression proves the read is still pending after 10 seconds and fails at 60 seconds; the client kills all 106 mutants.

File: src/shared/uptime-kuma/monitors.ts [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+      )
+    );
+  } catch {
+    // A malformed custom header belongs to a different, broken monitor.
+    return false;
+  }
+};
+
+const acceptsScheduledResponse = (statusCodes: string[]): boolean =>
+  statusCodes.includes("200-299") || statusCodes.includes("204");
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in b3692d6c. Accepted status codes are now evaluated as Kuma-style single values or inclusive ranges, so 200-399 and 100-599 reuse a monitor while 205-399 and 100-203 do not. Exact 204 and the existing 200-only negative case remain covered; monitor reconciliation kills all 111 mutants.

**Comment by stefan-burke** [comment]
The CodeRabbit README finding that GitHub could not attach inline is resolved in b3692d6c. The UPTIME_KUMA_URL row now explicitly requires CAN_BUILD_SITES=true together with the Kuma username and password. The pinned full precommit passes.

File: src/shared/uptime-kuma/monitors.ts [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+  authorization: string,
+  allowRaceDuplicates: boolean,
+): UptimeKumaMonitor | null => {
+  const matches = monitors.filter(
+    (monitor) =>
+      monitor.type === "http" &&
+      monitor.parent !== null &&
+      groupIds.includes(monitor.parent) &&
+      monitor.method === "POST" &&
+      monitor.url === url &&
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in f8a39afa. Matching now compares the URL origin plus the normalized path, so an existing `/scheduled/` monitor is reused instead of reported missing. Separate load and add regressions reproduce the trailing-slash case and prove that no duplicate is created; the monitor service kills all 114 mutants and full precommit passes.

File: src/shared/uptime-kuma/config.ts [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+  const raw =
+    getEnv("UPTIME_KUMA_INTERVAL_MINUTES") ??
+    String(DEFAULT_UPTIME_KUMA_INTERVAL_MINUTES);
+  const minutes = parsePositiveInt(raw);
+  if (minutes === null) {
+    throw new Error(
+      "UPTIME_KUMA_INTERVAL_MINUTES must be a positive whole number",
+    );
+  }
+  const seconds = minutes * 60;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/config.ts:54]: No upper cap is intended: operators may choose any positive whole-minute monitor frequency. Commit f8a39afa adds an explicit 60-minute configuration regression and corrects the README and maintenance guide so 15 minutes is described as the default, not a maximum. Invalid, fractional, zero, and overflowing values remain rejected; configuration kills all 59 mutants and full precommit passes.

File: src/shared/uptime-kuma/monitors.ts [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+const siteMonitor = (
+  monitors: UptimeKumaMonitor[],
+  groupIds: number[],
+  url: string,
+  authorization: string,
+  allowRaceDuplicates: boolean,
+): UptimeKumaMonitor | null => {
+  const matches = monitors.filter(
+    (monitor) =>
+      monitor.type === "http" &&
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts:115]: No active filter is intended because a paused monitor should be shown, not replaced by a duplicate that overrides the operator’s pause. The panel already renders `active: false` as “Paused”; f8a39afa adds a service regression proving a paused matching monitor, including its chosen interval, reaches that visible state. The operator can then resume it in Kuma. Full precommit passes.

File: src/shared/uptime-kuma/monitors.ts [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+      monitor.parent !== null &&
+      groupIds.includes(monitor.parent) &&
+      monitor.method === "POST" &&
+      monitor.url !== null &&
+      scheduledTarget(monitor.url) === target &&
+      acceptsScheduledResponse(monitor.acceptedStatusCodes) &&
+      hasAuthorization(monitor.headers, authorization),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in bd5e40c8. The Kuma response boundary now validates `authMethod` and the version-optional `bearer_token`, then reduces built-in bearer auth and custom headers to the one effective Authorization value Kuma sends. Custom Authorization still overrides built-in auth, unrelated headers preserve it, malformed headers remain a non-match, and raw auth fields do not spread into the service layer. Direct regressions cover each case and existing add/load tests prove the normalized authorization is reused without a duplicate; client mutation is 128/128, monitor mutation is 100/100, and full precommit passes.

File: src/shared/uptime-kuma/monitors.ts [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+  const matches = monitors.filter(
+    (monitor) =>
+      monitor.type === "http" &&
+      monitor.parent !== null &&
+      groupIds.includes(monitor.parent) &&
+      monitor.method === "POST" &&
+      monitor.url !== null &&
+      scheduledTarget(monitor.url) === target &&
+      acceptsScheduledResponse(monitor.acceptedStatusCodes) &&
+      monitor.authorization === authorization,
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts:123]: Resolved in 659e8464. The shared bearer parser now accepts the scheme case-insensitively and the Kuma response boundary rewrites valid custom values to the canonical `Bearer <token>` form before monitor matching. A direct client regression sends a lowercase custom bearer value and asserts the canonical effective authorization; exhaustive client mutation testing is 241/241.

File: src/shared/uptime-kuma/client.ts [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+  if (custom.authorization !== null) return custom.authorization;
+  return authMethod === "bearer" && bearerToken !== null
+    ? `Bearer ${bearerToken}`
+    : null;
+};
+
+const RawMonitorSchema = v.object({
+  accepted_statuscodes: v.array(v.string()),
+  active: ActiveSchema,
+  authMethod: v.string(),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts]: Resolved in 659e8464. Kuma monitor parsing now accepts `authMethod: null` and treats it as no built-in authorization, while keeping `bearer_token` nullable. A direct response-boundary regression loads an unauthenticated monitor with both fields null and proves the full monitor list remains readable; exhaustive client mutation testing is 241/241.

File: src/shared/uptime-kuma/client.ts [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+          list.received,
+        ]);
+        requireOk(response);
+        return Object.values(v.parse(MonitorListSchema, list.latest()));
+      } finally {
+        list.cancel();
+      }
+    },
+    login: async (username, password): Promise<void> => {
+      await version.read();
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts]: Resolved in 10d32174. Login now sends and validates the credential response first, then waits for the version-bearing post-login `info` event before returning. The regression starts with a version-hidden `info`, proves the login event is already sent, then supplies the supported version; separate cases prove bad credentials and two-factor login return without waiting. Exhaustive client mutation testing remains 241/241.

File: src/shared/uptime-kuma/client.ts [hunk 25]
```diff
@@ -0,0 +1,357 @@
...
+import type { UptimeKumaConfig } from "./config.ts";
+import {
+  createUptimeKumaSocket,
+  type UptimeKumaSocket,
+  uptimeKumaConnectionError,
+} from "./socket.ts";
+
+const SOCKET_TIMEOUT_MS = 10_000;
+const MONITOR_LIST_TIMEOUT_MS = 60_000;
+const UNSUPPORTED_VERSION_MESSAGE = "Uptime Kuma 2.4 or newer is required.";
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts]: Resolved in e8a06590. Locally generated client failures are now typed and mapped to built-sites catalog keys at the monitor service boundary, while messages returned by Kuma still pass through unchanged. Load regressions cover unsupported versions, two-factor login, version timeout, and monitor-list timeout; the add regression proves unsupported-version copy also reaches the flash through the catalog. Full precommit passes, with exhaustive mutation scores of 248/248 for the client and 192/192 for monitor reconciliation.

File: src/locales/en/built-sites.json [hunk 26]
```diff
@@ -26,6 +26,30 @@
   "built_sites.maintenance_provision": "Set up scheduled maintenance",
   "built_sites.maintenance_resend": "Send key to site again",
   "built_sites.maintenance_provisioned": "Scheduled maintenance key sent to the site.",
+  "built_sites.kuma_title": "Uptime Kuma monitor",
+  "built_sites.kuma_unconfigured": "Set the Uptime Kuma URL, username and password on this site to manage monitors here.",
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/locales/en/built-sites.json]: Resolved in ecd61573. The unconfigured panel now tells the owner to set the Uptime Kuma URL, username, and password on the builder, not on the child site. The rendered maintenance-panel regression asserts the builder wording, and full precommit passes.

File: src/shared/uptime-kuma/monitors.ts [hunk 27]
```diff
@@ -0,0 +1,327 @@
...
+};
+
+const kumaErrorMessage = (error: unknown): string =>
+  error instanceof UptimeKumaClientError
+    ? t(CLIENT_ERROR_MESSAGE_KEYS[error.kind])
+    : errorMessage(error);
+
+const errorMessage = (error: unknown): string => {
+  if (error instanceof Error) return error.message;
+  return "Uptime Kuma failed.";
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/monitors.ts]: Resolved in ec6e27db. The non-Error fallback now comes from `built_sites.kuma_failed` through `t(...)`. The regression enables `I18N_REPLACEMENTS=failed|stopped` and proves the add result becomes `Uptime Kuma stopped.`, so it would fail if the copy were hard-coded; monitor reconciliation remains at 192/192 mutation coverage and full precommit passes.

File: src/shared/uptime-kuma/socket.ts [hunk 28]
```diff
@@ -0,0 +1,265 @@
...
+      throw new Error("Uptime Kuma is already connected.");
+    this.#attempt(() => {
+      const raw = uptimeKumaWebSocketFactory.create(this.#url);
+      this.#raw = raw;
+      raw.onclose = () => this.#closed();
+      raw.onerror = () =>
+        this.#fail(new Error("Uptime Kuma connection failed."));
+      raw.onmessage = (event) => this.#message(event.data);
+      this.#connectTimer = setTimeout(
+        () => this.#fail(new Error("Uptime Kuma connection timed out.")),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/socket.ts]: Resolved in 03148432. Socket-generated connection, timeout, closure, request, and malformed-response failures now use shared `UptimeKumaError` kinds, and the monitor service maps every kind to built-sites catalog copy; messages returned by Kuma remain unchanged. Regressions cover socket protocol and lifecycle failures plus replacement-aware service rendering. Full precommit passes, with exhaustive mutation scores of 236/236 for the socket, 249/249 for the client, 210/210 for monitor reconciliation, 1/1 for the error type, 7/7 for the shared parser, and 10/10 for stored JSON.

File: src/shared/uptime-kuma/client.ts [hunk 29]
```diff
@@ -0,0 +1,373 @@
...
+        await call(socket, "login", {
+          password,
+          token: "",
+          username,
+        }),
+      );
+      if ("tokenRequired" in response) {
+        throw new UptimeKumaError("two_factor");
+      }
+      if (!response.ok) throw new Error(response.msg);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts]: Resolved in edc4a6b4. The login boundary now distinguishes Kuma 2.4’s translated `authIncorrectCreds` response from plain provider messages, converts it to the typed `incorrect_credentials` failure, and renders catalog copy telling the operator to check the builder credentials. The regression reproduces the exact `{ msg, msgi18n: true, ok: false }` payload; a separate case proves `msgi18n: false` still preserves Kuma’s plain message. Full precommit passes, with exhaustive mutation scores of 262/262 for the client, 212/212 for monitor reconciliation, and 1/1 for the shared error type.

File: scripts/mutation/equivalent-mutants.txt [hunk 30]
```diff
@@ -36,6 +36,28 @@ src/shared/forms/validation.ts:101:46  = → +=   # this assignment runs only wh
...
+src/shared/uptime-kuma/socket.ts:187:28  0 → -1   # the private map's missing set and retained empty set behave identically: once adds to either, while off and emit are no-ops for both
+src/shared/uptime-kuma/monitors.ts:94:12  === → ==   # firstById receives UptimeKumaMonitor|null and cannot receive undefined, so strict and loose equality agree when comparing it with null
+src/shared/uptime-kuma/monitors.ts:109:21  !== → !=   # monitor.parent is number|null and cannot be undefined, so strict and loose inequality agree when comparing it with null
+src/shared/uptime-kuma/monitors.ts:112:18  !== → !=   # monitor.url is string|null and cannot be undefined, so strict and loose inequality agree when comparing it with null
+src/shared/uptime-kuma/monitors.ts:133:15  !== → !=   # existing is UptimeKumaMonitor|null and cannot be undefined, so strict and loose inequality agree when comparing it with null
+src/shared/uptime-kuma/monitors.ts:188:13  === → ==   # config is UptimeKumaConfig|null and cannot be undefined, so strict and loose equality agree when comparing it with null
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/mutation/equivalent-mutants.txt]: Resolved in edc4a6b4. The stale `monitors.ts` and `client.ts` line/column keys were regenerated after the credential-translation refactor landed, and the latest exhaustive mutation run on `b42d43a` reports no ignore-list issues: client 265/265 killed (4 suppressed), monitors 234/234 killed (10 suppressed). A redundant `monitors.ts:298` entry whose `=== → ==` mutant is already killed by the `scheduledTaskKey: null` test was also removed in `b42d43a`.

File: src/shared/uptime-kuma/client.ts [hunk 31]
```diff
@@ -0,0 +1,373 @@
...
+    ? bearerAuthorization(bearerToken)
+    : null;
+};
+
+const RawMonitorSchema = v.object({
+  accepted_statuscodes: v.array(v.string()),
+  active: ActiveSchema,
+  authMethod: v.nullable(v.string()),
+  bearer_token: v.nullable(v.string()),
+  headers: v.nullable(v.string()),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts:99]: Resolved in b42d43a. The Kuma monitor schema now reads `upsideDown`, and a new `canCompleteScheduledRequest` guard rejects any monitor with `upsideDown: true` so the maintenance tab no longer treats an upside-down check as the site's scheduled monitor. A load regression reproduces an upside-down matching monitor and proves it is rejected; the boundary case (`upsideDown: false`) is accepted. Exhaustive client mutation is 265/265 (4 suppressed) and monitors 234/234 (10 suppressed).

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts:99]: Resolved in b42d43a. The schema now reads the `conditions` array, and `canCompleteScheduledRequest` rejects any monitor whose `conditions` array is non-empty, so an imported check with response-body or JSON assertions can no longer be treated as the site's scheduled monitor. A load regression reproduces a matching monitor with one condition and proves it is rejected; the empty-conditions case is accepted. Exhaustive client mutation is 265/265 (4 suppressed) and monitors 234/234 (10 suppressed).

File: src/shared/uptime-kuma/client.ts [hunk 32]
```diff
@@ -0,0 +1,373 @@
...
+  id: integerAtLeast(1),
+  interval: integerAtLeast(1),
+  method: v.string(),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts:102]: Resolved in b42d43a. The schema now reads the `timeout` field, and `canCompleteScheduledRequest` rejects any monitor whose timeout in milliseconds (`timeout * 1_000`) is below `MAINTENANCE_REQUEST_DEADLINE_MS` (25,000ms), so a manually-set 1-second timeout can no longer hide the add action behind a monitor that will go down whenever due work takes longer. A load regression reproduces a matching monitor with `timeout: 1` and proves it is rejected; a boundary test proves a monitor whose timeout in milliseconds equals the deadline is accepted. Exhaustive client mutation is 265/265 (4 suppressed) and monitors 234/234 (10 suppressed).

File: src/shared/uptime-kuma/client.ts [hunk 33]
```diff
@@ -0,0 +1,390 @@
...
+
+type CustomAuthorization = {
+  authorization: string | null;
+  valid: boolean;
+};
+
+const readCustomAuthorization = (
+  headers: string | null,
+): CustomAuthorization => {
+  if (headers === null) return { authorization: null, valid: true };
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/uptime-kuma/client.ts]: Resolved in 53e014a5. `readCustomAuthorization` now treats an empty `headers` string the same as `null` — no custom headers, so the built-in `bearer_token` is used — instead of feeding `""` into `JSON.parse`, which threw and marked the monitor invalid. A regression in `authorization.test.ts` reproduces a bearer-auth monitor with `headers: """"` and proves the effective authorization is `Bearer site key`; the test fails without the fix. Exhaustive client mutation remains 265/265 (4 suppressed).
